### PR TITLE
add todo_app-with-api

### PR DESCRIPTION
**Pull request by vneholiuk**
[DEMO LINK](https://vneholiuk.github.io/react_todo-app-with-api/)

### CODE REVIEW

File: src/App.tsx [hunk 1]
```diff
@@ -1,26 +1,304 @@
...
+
+    return () => {
+      clearTimeout(timeout);
+    };
+  }, [errorMessage]);
+
+  const activeTodos = todos.filter(todo => !todo.completed);
+  const completedTodos = todos.filter(todo => todo.completed);
+
+  let filterTodos = todos;
```
Comment (TarasHoliuk) [inline src/App.tsx:48]:
```suggestion
  let filteredTodos = todos;
```
filterTodos, with verb in name sounds like a function name

File: src/App.tsx [hunk 2]
```diff
@@ -1,26 +1,304 @@
...
+                  onToggle={() => handleToggleStatus(todo)}
+                  onRename={handleRename}
+                  isLoading={loadingTodos.includes(todo.id)}
+                />
+              ))}
+
+              {tempTodo && (
+                <TodoItem
+                  todo={tempTodo}
+                  isLoading={loadingTodos.includes(0)}
```
Comment (TarasHoliuk) [inline src/App.tsx:228]: we always want show loader for tempTodo
```suggestion
                  isLoading
```
So you can also avoid adding `0` to loadingTodos (BTW, there are only ids, not todos, so better name is `loadingTodoIds`)

File: src/components/TodoItem.tsx [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+}
+
+export const TodoItem: FC<Props> = ({
+  todo,
+  onDelete = () => {},
+  onToggle = () => {},
+  onRename = () => {},
+  isLoading,
+}) => {
+  const [isRemaining, setIsRemaining] = useState(false);
```
Comment (TarasHoliuk) [inline src/components/TodoItem.tsx:20]:
```suggestion
  const [isRenaming, setIsRenaming] = useState(false);
```
I guess you mean this

File: src/components/TodoItem.tsx [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+  todo: Todo;
+  onDelete?: () => void;
+  onToggle?: () => void;
+  onRename?: (todo: Todo) => void;
+  isLoading: boolean;
+}
+
+export const TodoItem: FC<Props> = ({
+  todo,
+  onDelete = () => {},
```
Comment (TarasHoliuk) [inline src/components/TodoItem.tsx:15]: you can do it optional and check before calling it

but also, you can improve the current approach: create some `noOperation` function (you can create better name)) above component and use it as default value for each one

File: src/App.tsx [hunk 5]
```diff
@@ -1,26 +1,304 @@
...
-        Copy all you need from the prev task:
-        <br />
-        <a href="https://github.com/mate-academy/react_todo-app-add-and-delete#react-todo-app-add-and-delete">
-          React Todo App - Add and Delete
-        </a>
-      </p>
-
-      <p className="subtitle">Styles are already copied</p>
-    </section>
+    <div className="todoapp">
```
Comment (TarasHoliuk) [inline src/App.tsx:182]: this component became quite big now
let's split it into a few separate components

**Review by TarasHoliuk — CHANGES_REQUESTED** [review]
Great job! Works perfect
Added a few small comments to refactor your code a bit

File: src/App.tsx [hunk 6]
```diff
@@ -1,26 +1,299 @@
...
+  const [status, setStatus] = useState<Status>('all');
+  const [title, setTitle] = useState('');
+  const [tempTodo, setTempTodo] = useState<Todo | null>(null);
+  const [loadingTodos, setLoadingTodos] = useState<number[]>([]);
+  const inputRef = useRef<HTMLInputElement>(null);
+
+  useEffect(() => {
+    inputRef.current?.focus();
+    getTodos()
+      .then((response: React.SetStateAction<Todo[]>) => setTodos(response))
```
Comment (volodymyr-soltys97) [inline src/App.tsx:29]:
```suggestion
      .then(setTodos)
```

File: src/App.tsx [hunk 7]
```diff
@@ -1,26 +1,299 @@
...
+            <footer className="todoapp__footer" data-cy="Footer">
+              <span className="todo__count" data-cy="TodosCounter">
+                {activeTodos.length} items left
+              </span>
+
+              <nav className="filter" data-cy="Filter">
+                <a
+                  href="#/"
+                  className={classNames('filter__link', {
+                    selected: status === 'all',
```
Comment (volodymyr-soltys97) [inline src/App.tsx:237]: Create a enum for 'all', 'active', 'completed' and use it everywhere

**Review by volodymyr-soltys97 — CHANGES_REQUESTED** [review]
Good job 👍 
Let's make your code better

**Review by volodymyr-soltys97 — APPROVED** [review]
Well done! 🔥

**Pull request closed without merging**